File: app/src/main/java/com/slidely/photoframe/repository/RepositoryProvider.java
package com.slidely.photoframe.repository;

import android.app.Activity;
import android.support.v7.app.AppCompatActivity;

public final class RepositoryProvider {
    private static MediaRepository mediaRepository;
    private static PreferenceRepository preferenceRepository;

    public static MediaRepository provideMediaRepository(AppCompatActivity activity) {
        if (mediaRepository == null)
            mediaRepository = new DefaultMediaRepository(activity);
        return mediaRepository;
    }

    public static PreferenceRepository providePreferenceRepository(Activity activity) {
        if (preferenceRepository == null)
            preferenceRepository = new DefaultPreferenceRepository(activity);
        return preferenceRepository;
    }
}

File: app/src/main/java/com/k4dima/photoframe/features/main/data/di/RepositoryProvider.kt
package com.k4dima.photoframe.features.main.data.di

import android.content.Context
import com.k4dima.photoframe.features.main.data.DefaultMediaRepository
import com.k4dima.photoframe.features.main.data.DefaultPreferenceRepository
import com.k4dima.photoframe.features.main.data.MediaRepository
import com.k4dima.photoframe.features.main.data.PreferenceRepository
import com.k4dima.photoframe.features.main.ui.MainActivity
import com.k4dima.photoframe.features.main.ui.di.MainScope
import com.squareup.sqlbrite3.SqlBrite
import dagger.Binds
import dagger.Module
import dagger.Provides

@Module
abstract class RepositoryProvider {
    @Module
    companion object {
        @MainScope
        @JvmStatic
        @Provides
        fun sqlBrite() = SqlBrite.Builder().build()

        @JvmStatic
        @Provides
        fun contentResolver(mainActivity: MainActivity) = mainActivity.contentResolver!!

        @JvmStatic
        @Provides
        fun context(mainActivity: MainActivity): Context = mainActivity.applicationContext
    }

    @MainScope
    @Binds
    abstract fun provideMediaRepository(defaultMediaRepository: DefaultMediaRepository): MediaRepository

    @MainScope
    @Binds
    abstract fun providePreferenceRepository(defaultPreferenceRepository: DefaultPreferenceRepository): PreferenceRepository
}